share/eds/p2p/server.go

```diff
@@ -0,0 +1,161 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	p2p_pb "github.com/celestiaorg/celestia-node/share/eds/p2p/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+)
+
+const (
+	// writeDeadline sets timeout for sending messages to the stream
+	writeDeadline = time.Second * 5
+	// readDeadline sets timeout for reading messages from the stream
+	readDeadline = time.Minute
+)
+
+// Server is responsible for serving ODSs for blocksync over the ShrEx/EDS protocol.
+type Server struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	host       host.Host
+	protocolID protocol.ID
+
+	store *eds.Store
+}
+
+// NewServer creates a new ShrEx/EDS server.
+func NewServer(host host.Host, store *eds.Store, protocolSuffix string) *Server {
+	return &Server{
+		host:       host,
+		store:      store,
+		protocolID: protocolID(protocolSuffix),
+	}
+}
+
+func (s *Server) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.host.SetStreamHandler(s.protocolID, s.handleStream)
+	return nil
+}
+
+func (s *Server) Stop(context.Context) error {
+	defer s.cancel()
+	s.host.RemoveStreamHandler(s.protocolID)
+	return nil
+}
+
+func (s *Server) handleStream(stream network.Stream) {
+	log.Debug("server: handling eds request")
+
+	// read request from stream to get the dataHash for store lookup
+	req, err := s.readRequest(stream)
+	if err != nil {
+		log.Errorw("server: reading request from stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	// ensure the requested dataHash is a valid root
+	err = share.DataHash(req.Hash).Validate()
```

Comment:
i'd cast into a datahash, call validate, and then pass as a datahash to the getcar method

like: 
hash := share.DataHash(req.Hash)
err := hash.Validate()
if err blah blah

GetCar(..., hash, ..)